"nvim -es": disable shada

justinmk opened the pull request:
# Problem

`nvim -es` (and `nvim -Es`) is the recommended way to non-interactively run commands/vimscript. But it enables shada by default, which is usually not wanted.

# Solution

- Disable shada by default for `nvim -es/-Es`. This can be overridden by `-i foo` if needed.
- Do NOT change the 'loadplugins' default. See below...

Why enable config+packages by default:

- User config + packages _should_ be enabled by default, for both `nvim -es` and `nvim -l`. Else any Lua packages you have can't be accessed without `-u path/to/config`, which is clumsy.
    - Use-cases:
      ```
      nvim --headless "+Lazy! sync" +qa
          would become: nvim -es "+Lazy! sync"
      nvim --headless +PlugInstall +qall
          would become: nvim -es +PlugInstall
      ```
- Opt-out (`--clean` or `-u NONE`) is much easier than opt-in (`-u path/to/config`).
-  User config/packages are analogous to pip packages, which are expected when doing `python -c ...`.

folke commented:
@justinmk probably just an example, but for Lazy that's not a problem. Lazy doesn't use packages, manages the rtp itself, and has a custom lua loader.

When starting up, it checks if `loadplugins=true`. If that's the case, it sets `loadplugins=false`, so that the default plugins don't do anything.  If the users disables plugins with `loadplugins=false`, then it doesn't load the plugins.

Either way, I don't think it's a good idea to disable plugins for `-es`? What if the user wants to run some lua or vimscript code that depends on a plugin?

With your suggested change, `--headless` would still work like before?

justinmk commented:
> Either way, I don't think it's a good idea to disable plugins for `-es`? What if the user wants to run some lua or vimscript code that depends on a plugin?

yeah, see "Open questions" above. I'm thinking about "inversing" this PR. Then commands like `:Lazy` would be usable with `nvim -es`.

> With your suggested change, `--headless` would still work like before?

`--headless` won't be affected

lewis6991 commented:
> opt-out (-u NONE) is much easier than opt-in -u path/to/config +"set rtp=..."

I think opt-out is much better here.

justinmk committed:
feat(startup): -es/-Es disable shada

Problem:
`nvim -es` (and `nvim -Es`) is the recommended way to non-interactively
run commands/vimscript. But it enables shada by default, which is
usually not wanted.

Solution:
- Disable shada by default for `nvim -es/-Es`. This can be overridden by
  `-i foo` if needed.
- Do NOT change the 'loadplugins' default.
  - User config + packages _should_ be enabled by default, for both `nvim
    -es` and `nvim -l`. Else any Lua packages you have can't be accessed
    without `-u path/to/config`, which is clumsy.
      - Use-cases:
        ```
        nvim --headless "+Lazy! sync" +qa
            would become: nvim -es "+Lazy! sync"
        nvim --headless +PlugInstall +qall
            would become: nvim -es +PlugInstall
        ```
  - Opt-out (`--clean` or `-u NONE`) is much easier than opt-in (`-u
    path/to/config`).
  -  User config/packages are analogous to pip packages, which are
     expected when doing `python -c ...`.

related: 7c94bcd2d77e2e54b8836ab8325460a367b79eae
related: ddd0eb6f5120a09b97867d2561ea61309038ccd2

justinmk commented:
Refocused this PR to only disable *shada* by default.

> > opt-out (-u NONE) is much easier than opt-in -u path/to/config +"set rtp=..."
> 
> I think opt-out is much better here.

Agree. Unfortunately that was already shipped for `nvim -l` https://github.com/neovim/neovim/commit/f43de742e881e54a3602e00c8c247cecca65a266 . I'm in favor of revisiting that; tracked in #31878.
